apps/hash-ai-worker-ts/src/workflows/run-flow-workflow.ts

```diff
@@ -38,18 +39,29 @@ const log = (message: string) => {
   console.log(message);
 };
 
-const doesFlowStepHaveSatisfiedDependencies = (step: FlowStep) => {
+const doesFlowStepHaveSatisfiedDependencies = (params: {
+  step: FlowStep;
+  flowDefinition: FlowDefinition;
+}) => {
+  const { step, flowDefinition } = params;
+
   if (step.kind === "action") {
     /**
-     * An action step has satisfied dependencies if all of its required inputs
-     * have been provided.
+     * An action step has satisfied dependencies if all of its inputs have
+     * been provided, based on the input sources defined in the step's
+     * definition.
+     *
+     * We don't need to check if all required inputs have been provided,
+     * as this will have been enforced when the flow was validated.
      */
-    const requiredInputs = actionDefinitions[
-      step.actionDefinitionId
-    ].inputs.filter((input) => input.required);
 
-    return requiredInputs.every((requiredInput) =>
-      step.inputs?.some(({ inputName }) => requiredInput.name === inputName),
+    const { inputSources } = getStepDefinitionFromFlowDefinition({
+      step,
+      flowDefinition,
+    });
+
+    return inputSources.every((inputSource) =>
+      step.inputs?.some(({ inputName }) => inputSource.inputName === inputName),
     );
```

Comment:
This change breaks the ability to have a step with optional inputs be fed from a step that has optional outputs, because the outputs may be absent.

In my branch I've changed it to this:
```suggestion
    const actionDefinition = actionDefinitions[step.actionDefinitionId];

    return inputSources.every((inputSource) => {
      const inputDef = actionDefinition.inputs.find(
        (def) => def.name === inputSource.inputName,
      );
      if (!inputDef) {
        throw new Error(
          `Definition for inputName '${inputSource.inputName}' in step ${step.stepId} not found in action definition ${step.actionDefinitionId}`,
        );
      }
      if (!inputDef.required) {
        return true;
      }

      return step.inputs?.some(
        ({ inputName }) => inputSource.inputName === inputName,
      );
    });
```

I may be misunderstanding how this is supposed to work. Here's the query I encountered it in, where the trigger doesn't output some optional configuration settings, which the step it's feeding doesn't require, but the current code throws an error.
```graphql
mutation {
  startFlow(
    flowTrigger: {
      triggerDefinitionId: "userTrigger"
      outputs: [
        {
          outputName: "url",
          payload: { 
            value: "https://www.simplilearn.com/ice9/free_resources_article_thumb/what_is_image_Processing.jpg"
          }
        }
      ]
    }
    flowDefinition: {
      name: "Save File From Url"
      flowDefinitionId: "saveFileFromUrl"
      trigger: {
        triggerDefinitionId: "userTrigger"
        kind: "trigger"
        outputs: [
          { payloadKind: "Text", name: "url", array: false, required: true }
          { payloadKind: "Text", name: "description", array: false, required: false }
          { payloadKind: "Text", name: "displayName", array: false, required:false }
        ]
      }
      steps: [
        {
          stepId: "1"
          kind: "action"
          actionDefinitionId: "getFileFromUrl"
          inputSources: [
            {
              inputName: "url"
              kind: "step-output"
              sourceStepId: "trigger"
              sourceStepOutputName: "url"
            }
            {
              inputName: "description"
              kind: "step-output"
              sourceStepId: "trigger"
              sourceStepOutputName: "description"
            }
            {
              inputName: "displayName"
              kind: "step-output"
              sourceStepId: "trigger"
              sourceStepOutputName: "displayName"
            }
          ]
        }
      ]
      outputs: [
        {
          stepId: "1"
          stepOutputName: "fileEntity"
          name: "fileEntity"
          payloadKind: "Entity"
          array: false
        }
      ]
    }
  )
}
```

Reply:
Thanks for catching this Ciaran, something else that needs to be taken into consideration here is that we don't want steps to be processed, if they have an optional input source pointing another step that hasn't been processed.

I think I've addressed this in https://github.com/hashintel/hash/pull/4305/commits/b4f67375189ab3e595f90f4019cd0babf03eebd9, have provided a few comments to explain.

For context, this function is important because it determines whether a step is run. So it should only return `true`, if all the required inputs are provided, and any optional inputs from other runnable steps has been provided.

What I'm still a bit unsure about is how to handle optional inputs with a `parallel-group-input` source. I've made it so that they wait for the parallel group to pass this input, even if it's optional, otherwise the parallel group won't parallelise the child nodes.

Reply:
Ahh, got it – the explanation and the comments are very helpful, thank you